lib/school_of_fish.rb

```diff
@@ -24,22 +35,81 @@ def update(dt)
     elsif (rand<0.01)
       @y_fish_speed = 1*rand-0.5
     end
+
     @x_position += @x_fish_speed
     @y_position += @y_fish_speed
     @rotation += dt*@rotation_speed
+
+    update_dispersing(dt)
   end
 
+
   def draw
+      color = (
+      @color_a.floor * (2**(3*8)) +
+      @color_r.floor * (2**(2*8)) +
+      @color_g.floor * (2**(1*8)) +
+      @color_b.floor * (2**(0*8))
+    )
+
     @image_tiles.each.with_index do |tile, tile_index|
       tile.draw_rot(
         @x_position,
         @y_position,
         @z_position,
         (tile_index % 2 == 0 ? @rotation : -1 * @rotation) / (1 + tile_index),
-        0.5,0.5,1,1,
-        0x50_000000
+        @school_density,@school_density,@school_size,@school_size,
+        color
       )
     end
   end
 
+  private
+
+    def update_dispersing(dt)
+      if(@dispersing)
+        decrease_density(dt)
+        shrink(dt)
+        fade_out(dt)
+        bluer(dt)
+        check_if_done_dispersing
+      else
+        distance = check_distance_to(@player)
+        @dispersing = distance < 100
+      end
+    end
+    
+    def check_distance_to(object)
+      distance = Math.sqrt((object.y_position-@y_position)**2 + (object.x_position-@x_position)**2)
+      distance
+    end
+
+    def shrink(dt)
+      @school_size += -dt*(1/10000.to_f)
+    end
+
+    def decrease_density(dt)
+      @school_density += -dt*(0.5/700.to_f)
+    end
+
+    def fade_out(dt)
+      if (@color_a > 0)
+        @color_a = @color_a - dt*(255/2000.to_f)
+      end
+    end
+
+    def bluer(dt)
+      if (@color_b < 255)
+        @color_b = @color_b + dt*(255/500.to_f)
+      end
+      
+    end
+
+    def check_if_done_dispersing
+      if (@color_a<0 || @color_b>255)
+        @dispersing = false
+        initialize(rand(640), rand(480), 50, @player)
```

Comment:
no

Reply:
This is basically just teleporting, and actually it might be a good solution if we "sleep" for a duration of time (disable drawing, exit `update` early, etc.) before *rematerializing* in a random location. Alternatively, you *can* make it look like the fish are actually moving to the new location. For example, they fade out and back in while growing and then shrinking all while moving straight to the new position.

This would leave us in a position where we instantiate only one `SchoolOf Fish` for the entire scene. We still havent approached the issue of how we are going to *destroy* entities. With only one instance it may look like we circumvent the problem of *reinstantiation* but I think it's a code smell. I mean, we are going to have to instantiate and destroy things eventually.

If it were up to me, I'd create a `SchoolOfFishController` class that inherits from another new class called `ObjectController` (or something like that). The ObjectController class would have interfaces that other objects like `Window`, `Player`, etc. can interact with directly. It could sort of be a wrapper for game objects in general but that's another discussion...

Reply:
you can move forward with this strategy if you make an alteration for me: instead of calling `initialize`, make a new method `set_position` or something. just use that method here and inside `initialize`